src/main/java/net/fabricmc/fabric/mixin/registry/MixinIdRegistry.java

```diff
@@ -79,87 +79,120 @@ public void setPre(int id, Identifier identifier, Object object, CallbackInfoRet
 			for (Identifier id : registry.getIds()) {
 				//noinspection unchecked
-				initialIdMap.put(id, registry.getRawId(registry.get(id)));
+				fabric_prevIndexedEntries.put(id, registry.getRawId(registry.get(id)));
 			}
 		}
 
-		if (reallocateMissingEntries) {
+		// If we're AUTHORITATIVE, we append entries which only exist on the
+		// local side to the new entry list. For REMOTE, we instead drop them.
+		if (mode == RemapMode.AUTHORITATIVE) {
 			int maxValue = 0;
 
-			Object2IntMap<Identifier> idMapOld = idMap;
-			idMap = new Object2IntOpenHashMap<>();
-			for (Identifier id : idMapOld.keySet()) {
-				int v = idMapOld.getInt(id);
-				idMap.put(id, v);
+			Object2IntMap<Identifier> oldRemoteIndexedEntries = remoteIndexedEntries;
+			remoteIndexedEntries = new Object2IntOpenHashMap<>();
+			for (Identifier id : oldRemoteIndexedEntries.keySet()) {
+				int v = oldRemoteIndexedEntries.getInt(id);
+				remoteIndexedEntries.put(id, v);
 				if (v > maxValue) maxValue = v;
 			}
 
 			for (Identifier id : registry.getIds()) {
-				if (!idMap.containsKey(id)) {
+				if (!remoteIndexedEntries.containsKey(id)) {
 					LOGGER.warn("Adding " + id + " to registry.");
-					idMap.put(id, ++maxValue);
+					remoteIndexedEntries.put(id, ++maxValue);
 				}
 			}
+		} else if (mode == RemapMode.REMOTE) {
+			// TODO: Is this what mods really want?
+			Set<Identifier> droppedIds = new HashSet<>();
+
+			for (Identifier id : registry.getIds()) {
+				if (!remoteIndexedEntries.containsKey(id)) {
+					droppedIds.add(id);
+				}
+			}
+
+			entries.keySet().removeAll(droppedIds);
 		}
 
-		if (listeners != null) {
-			for (RegistryListener listener : listeners) {
+		// Inform about registry clearing.
+		if (fabric_listeners != null) {
+			for (RegistryListener listener : fabric_listeners) {
 				listener.beforeRegistryCleared(registry);
 			}
 		}
 
-		// We don't really need to clear anything but indexedEntries yet.
+		// entries was handled above, if it was necessary.
 		indexedEntries.clear();
 		nextId = 0;
 
-		List<Identifier> idsInOrder = new ArrayList<>(idMap.keySet());
-		idsInOrder.sort(Comparator.comparingInt(idMap::getInt));
+		List<Identifier> orderedRemoteEntries = new ArrayList<>(remoteIndexedEntries.keySet());
+		orderedRemoteEntries.sort(Comparator.comparingInt(remoteIndexedEntries::getInt));
 
-		for (Identifier identifier : idsInOrder) {
-			int id = idMap.getInt(identifier);
+		for (Identifier identifier : orderedRemoteEntries) {
+			int id = remoteIndexedEntries.getInt(identifier);
 			T object = entries.get(identifier);
+
+			// Warn if an object is missing from the local registry.
+			// This should only happen in AUTHORITATIVE mode, and as such we
+			// throw an exception otherwise.
 			if (object == null) {
-				LOGGER.warn(identifier + " missing from registry, but requested!");
+				if (mode != RemapMode.AUTHORITATIVE) {
+					throw new RemapException(identifier + " missing from registry, but requested!");
```

Comment:
Does this mean that trying to connect to a server that has blocks/items/etc. that the client doesn't will throw?

Reply:
Yes.